### PR TITLE
doc/dev: testing guidelines update for GN-105

### PR DESCRIPTION
**Pull request by bobheadxi**
Incorporates changes from https://github.com/sourcegraph/handbook/pull/2055 into the testing principles page, which is now testing principles and guidelines.

Closes https://github.com/sourcegraph/sourcegraph/issues/29894 , closes https://github.com/sourcegraph/sourcegraph/issues/29764 , closes https://github.com/sourcegraph/sourcegraph/issues/29889

Enforcement: https://github.com/sourcegraph/sourcegraph/pull/30278

### CODE REVIEW

File: doc/dev/background-information/testing_principles.md [hunk 1]
```diff
@@ -1,55 +1,39 @@
...
-**We do not tolerate flaky tests of any kind.** Any engineer that sees a flaky test in [continuous integration](./continuous_integration.md) should immediately [disable the flaky test](continuous_integration.md#flaky-tests).
-
-Why are flaky tests undesirable? Because these tests stop being an informative signal that the engineering team can rely on, and if we keep them around then we eventually train ourselves to ignore them and become blind to their results. This can hide real problems under the cover of flakiness.
-
-Other kinds of flakes include [flaky steps](continuous_integration.md#flaky-steps) and [flaky infrastructure](continuous_integration.md#laky-infrastructure)
-
-## Testing pyramid
+The testing pyramid is a helpful way to determine the most appropriate type of test when deciding how to test a change:
 
 ![Testing pyramid](testing-pyramid.svg)
```
Comment (jhchabran) [inline doc/dev/background-information/testing_principles.md:35]: What about adding some explanation about the pyramid? 

```suggestion
The closer a test is to the bottom, the larger the scope of that test is. It means that failures will be harder to link to the actual cause. Tests at the bottom are notoriously slower than at the top. 

It's important to take into account this trade-off when deciding at which level to implement a test. Please refer to each testing level below for more details.
```

File: doc/dev/background-information/testing_principles.md [hunk 2]
```diff
@@ -101,19 +85,52 @@ Examples:
...
+The list below designates source code exempt from the testing guidelines because they do not directly impact the behaviour of the application in any way.
+
+- [sourcegraph/sourcegraph](https://github.com/sourcegraph/sourcegraph)
+  - `dev/*`: internal tools, scripts for the local environment and continuous integration.
+  - `enterprise/dev/*`: internal tools, scripts for the local environment and continuous integration that fall under the [Sourcegraph Enterprise license](https://github.com/sourcegraph/sourcegraph/blob/main/LICENSE.enterprise).
+
 ## Ownership
 
 - [DevX Team](https://handbook.sourcegraph.com/engineering/enablement/dev-experience) owns build and test infrastructure.
 - [Frontend Platform Team](https://handbook.sourcegraph.com/engineering/enablement/frontend-platform) owns any tests that are driven through the browser.
```
Comment (jhchabran) [inline doc/dev/background-information/testing_principles.md:136]: Minor clarification

```suggestion
- [Frontend Platform Team](https://handbook.sourcegraph.com/engineering/enablement/frontend-platform) owns any tests that are driven through the browser.
- All other tests are owned by the teams responsible for the domain being tested. 
```

**Review by jhchabran — APPROVED** [review]
We're getting closer to the finish line 🚀

**Comment by bobheadxi** [comment]
I'm pulling the PR template updates from https://github.com/sourcegraph/sourcegraph/pull/30278 into this PR

**Pull request merged**